src/dev/riscv/HiFive.py

```diff
@@ -129,19 +138,25 @@ def _off_chip_ranges(self):
             for dev in self._off_chip_devices()
         ]
 
-    def attachOnChipIO(self, bus):
-        """Attach on-chip IO devices, needs modification
-        to support DMA
-        """
-        for device in self._on_chip_devices():
-            device.pio = bus.mem_side_ports
-
-    def attachOffChipIO(self, bus):
-        """Attach off-chip IO devices, needs modification
-        to support DMA
-        """
-        for device in self._off_chip_devices():
-            device.pio = bus.mem_side_ports
+    def attachOnChipIO(self, bus, ruby=False, dma_ports=None):
```

Comment:
Likewise, it would be nice to start proving typing hints some of these functions as they grow in complexity (again, not a hard requirement for this PR, but it would be nice to have):

```suggestion
    def attachOnChipIO(self, bus : SystemXBar, ruby : bool=False, dma_ports : Optional[Port]=None ) -> None:
```

Reply:
Done